torchtitan/parallelisms/parallelize_llama.py

```diff
@@ -129,15 +145,281 @@ def get_tp_parallel_strategy(
+    """Helper for splitting ordered list of layer names into layers per stage.
+
+    split_points is a list of layer names, each layer will be the first layer in a stage
+    """
+    stages = []
+    cur = []
+
+    for name in fqns:
+        if name in split_points:
+            assert len(
+                cur
+            ), f"{name} is not a valid split point, do not specify the first layer of stage 0"
+            stages.append(cur)
+            cur = []
+        cur.append(name)
+
+    stages.append(cur)
+    print(f"Split using points {split_points}, got statges {stages}")
+    return stages[stage_id]
+
+
+def apply_pipeline_parallelism_manual(
+    model, world_mesh, parallel_dims, job_config: JobConfig, device, model_config: Dict
+):
+    """
+    This API gets individual torch.nn.Module objects for each pipeline stage (including virtual stages).
+
+    The SPMD parallelisms should be applied to
+    """
+    pp_mesh = world_mesh["pp"]
+    pp_rank = pp_mesh.get_local_rank()
+    pp_size = pp_mesh.size()
+    # heuristically == PP dim but should be a config
+    microbatches = parallel_dims.pp
+    stage_idx = pp_rank  # TODO support virtual stages
+    layers_per_rank = len(model.layers) // parallel_dims.pp
+    layer_offset = layers_per_rank * pp_rank
+    this_stage_layer_names = [
+        f"layers.{i + layer_offset}" for i in range(layers_per_rank)
+    ]
+    if pp_rank == 0:
+        this_stage_layer_names.insert(0, "tok_embeddings")
+        assert "layers.0" in this_stage_layer_names
+    elif pp_rank == pp_size - 1:
+        this_stage_layer_names.append("norm")
+        this_stage_layer_names.append("output")
+        assert "layers.1" in this_stage_layer_names
+
+    fqns = _llama_fqns(len(model.layers))
+    new_names = split_stage_fqns(
+        fqns, job_config.experimental.pipeline_parallel_split_points, pp_rank
+    )
+    assert len(new_names) == len(this_stage_layer_names), (
+        len(new_names),
+        len(this_stage_layer_names),
+    )
+    for n, n_ in zip(new_names, this_stage_layer_names):
+        assert n == n_, (n, n_)
+
+    input_seqlen = 2048  # TODO hack
+
+    model = TransformerChunk(model, this_stage_layer_names, device, input_seqlen)
+    # Create a pipeline representation from the model
+
+    # TODO(whc) once ManualPipelineStage supports lazy shape inference, we can leave model on meta device longer and
+    # get rid of the input shape hardcoded here. For now, it should not be a big deal since we only materialize the
+    # layers of the model that map to this stage, not the whole model.
+
+    # Get example input
+    if pp_rank == 0:
+        input_shape = (job_config.training.batch_size, job_config.training.seq_len)
+        input = torch.randint(
+            model_config.vocab_size, input_shape, dtype=torch.int64, device=device
+        )
+
+        # HACK- can't use shape inference via execution of the PP stage inside ManualPipelineStage API, becuase the
```

Comment:
proposal to land this part as-is. We need to add the lazy shape inference part to manual stage api, but i think we can clean this up after it lands.